core/src/main/java/com/linecorp/armeria/client/cookie/DefaultCookieJar.java

```diff
@@ -0,0 +1,157 @@
+
+import com.google.common.base.Strings;
+
+import com.linecorp.armeria.common.Cookie;
+import com.linecorp.armeria.common.Cookies;
+
+/**
+ * A default in-memory {@link CookieJar} implementation.
+ */
+final class DefaultCookieJar implements CookieJar {
+
+    private final Set<Cookie> store;
+    /**
+     * Used to find cookies for a host that matches a domain. For example, if there is a domain example.com,
+     * host example.com or foo.example.com will match all cookies in that entry.
+     */
+    private final Map<String, Set<Cookie>> filter;
+    private final CookiePolicy cookiePolicy;
+    private final ReentrantLock lock;
+
+    DefaultCookieJar() {
+        this(CookiePolicy.ACCEPT_ORIGINAL_SERVER);
+    }
+
+    DefaultCookieJar(CookiePolicy cookiePolicy) {
+        this.cookiePolicy = cookiePolicy;
+        store = new HashSet<>();
+        filter = new HashMap<>();
+        lock = new ReentrantLock();
+    }
+
+    @Override
+    public Cookies get(URI uri) {
+        requireNonNull(uri, "uri");
+        if (store.isEmpty()) {
+            return Cookies.of();
+        }
+        final String host = uri.getHost();
+        final String path = Strings.isNullOrEmpty(uri.getPath()) ? "/" : uri.getPath();
+        final boolean secure = "https".equalsIgnoreCase(uri.getScheme());
+        final Set<Cookie> cookies = new HashSet<>();
+        lock.lock();
+        try {
+            filterGet(cookies, host, path, secure);
+        } finally {
+            lock.unlock();
+        }
+        return Cookies.of(cookies);
+    }
+
+    @Override
+    public void set(URI uri, Cookies cookies) {
+        requireNonNull(uri, "uri");
+        requireNonNull(cookies, "cookies");
+        lock.lock();
+        try {
+            for (Cookie cookie : cookies) {
+                // if domain is not specified, set the domain as the original request host
+                // https://tools.ietf.org/html/rfc6265#section-5.3
+                if (cookie.domain() == null) {
+                    cookie = cookie.toBuilder().domain(uri.getHost()).hostOnly(true).build();
+                }
+                // if path is not specified or the first character is not '/', set the default path
+                // https://tools.ietf.org/html/rfc6265#section-5.2.4
+                // https://tools.ietf.org/html/rfc6265#section-5.1.4
+                final String cookiePath = cookie.path();
+                if (Strings.isNullOrEmpty(cookiePath) || cookiePath.charAt(0) != '/') {
+                    String path = uri.getPath();
+                    if (Strings.isNullOrEmpty(path) || path.charAt(0) != '/') {
+                        path = "/";
+                    } else {
+                        final int i = path.lastIndexOf('/');
+                        if (i > 0) {
+                            path = path.substring(0, i);
+                        }
+                    }
+                    cookie = cookie.toBuilder().path(path).build();
```

Comment:
Could you extract this block as a method with `@VisibleForTest` annotation and add test cases for the path normalization?

Reply:
I added the test in the latest commit, please let me know if it satisfies. It covers the above discussion as well https://github.com/line/armeria/pull/3118#discussion_r515101848.